[API] Fix class loading issues during development

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19190
## Description

This is a new try for class loading in APIv3 classes, we either:
- FULLY qualify a class and trust the rails autoloader
- qualify as few as possible (see below) and require the corresponding class

That means we do not (yet) completely give up on the rails autoloader, but we circumvent a fair amount of its magic (and problems).

The hope is to eliminate the following problems in development mode:
- NameErrors because the autoloader infers the wrong prefix
- "object is not missing constant" errors occuring more rarely

What we still want to maintain is:
- The ability to automatically resolve a class' file when fully qualifying it
- The automatic reloading of changed classes during development

So far it looks as if this is still working. Lets see if that is true ^^
## What does "qualify as few as possible" mean?

Ideally, if we already require a class we do not want to qualify at all:

``` ruby
require 'api/v3/statuses/status_representer'

# in api/v3/statuses/status_api.rb
module API
  module V3
    module Statuses
      # ... [insert actual code] ...
      StatusRepresenter
    end
  end
end
```

However, when we don't share the same module we are forced to go up to the first common module, for Ruby to be able to perform the lookup successfully:

``` ruby
require 'api/v3/statuses/status_representer'

# in api/v3/whatever/another_api.rb
module API
  module V3
    module Whatever
      # this will work, our common module is API::V3, so we can refer to Statuses safely
      Statuses::StatusRepresenter

      # this will FAIL, because there is neither of the following:
      #  - ::API::V3::Whatever::StatusRepresenter
      #  - ::API::V3::StatusRepresenter
      #  - ::API::StatusRepresenter
      #  - ::StatusRepresenter
      StatusRepresenter
    end
  end
end
```

Note that Ruby will take the first match. So if you refer to `Utilities::Foo` from `A::B::C` and there are both, `A::Utilities` and `A::B::Utilities` it will take the latter, because it starts from your module going to the root. That means if `Foo` is defined in `A::Utilities` it will **not** be found.

Those are ruby basics, however I was able to not know them (I blame rails for taking too much work from me ^^)... So I thought I better repeat them here.
## Affected files

This PR only considers the actual endpoint classes, i.e. files ending in `_api.rb`. This should be a good start and to my knowledge was the source of the most errors recently.

If we choose that this is the path we want to go, **_all files**_ under `lib/api/` should follow that pattern. It should not be neccessary to touch specs, because they only run in the `test` environment.

NobodysNightmare committed:
specify classes explicitly rather than implicitly

This is a new try for the APIv3 classes, we either:
- FULLY qualify a class and let the autoloader do the rest
- don't qualify at all, but require the class explicitly

Thus we do not (yet) completely give up on the rails autoloader,
but we circumvent a fair amount of its magic (and problems).

The hope is to eliminate the following problems in development mode:
- NameErrors because the autoloader infers the wrong prefix
- "object is not missing constant" errors occuring more rarely

Let's see how that turns out to work...

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-api-load-issues.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-api-load-issues.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare commented:
Apparently this does not work at all -.-

I hate class loading/resolution...

NobodysNightmare committed:
use partial namespace prefixes where neccessary

apparently using Rails leads to people (like me) not learning how Ruby actually looks up Constants.

We need to provide the prefix whenever the Constant can't be found in our scope or a parent scope.

NobodysNightmare committed:
style

NobodysNightmare committed:
more namespace qualification

NobodysNightmare committed:
add another forgotten prefix

now I should have found them all *knock on wood*

ulferts commented:
The approach looks reasonable. Only time will tell wether it works, though, because the autoloader won't tell anything.

Needs rebase/merge @NobodysNightmare
